worker/caasoperator/caasoperator.go

```diff
@@ -255,17 +261,48 @@ func (op *caasOperator) loop() (err error) {
 		return errors.Trace(err)
 	}
 
-	aliveUnits := make(set.Strings)
+	// Keep a record of the alive units an a channel used to notify
+	// their uniter workers when the charm version has changed.
+	aliveUnits := make(map[string]chan struct{})
 
 	if err = restartWatcher(); err != nil {
 		err = errors.Annotate(err, "(re)starting watcher")
 		return errors.Trace(err)
 	}
 
+	// We should not do anything until there has been a change
+	// to the remote state. The watcher will trigger at least
+	// once initially.
+	select {
+	case <-op.catacomb.Dying():
+		return op.catacomb.ErrDying()
+	case <-watcher.RemoteStateChanged():
+	}
+
 	for {
 		select {
 		case <-op.catacomb.Dying():
 			return op.catacomb.ErrDying()
+		case <-watcher.RemoteStateChanged():
+			snap := watcher.Snapshot()
+			if charmModified(localState, snap) {
+				// Charm changed so download and install the ne version.
+				charmURL, charmModifiedVersion, err := op.ensureCharm()
+				if err != nil {
+					return errors.Annotatef(err, "error downloading updated charm %v", charmURL.String())
+				}
+				localState.CharmURL = charmURL
+				localState.CharmModifiedVersion = charmModifiedVersion
+				// Notify all uniters of the change so they run the upgrade-charm hook.
+				for unitId, changedChan := range aliveUnits {
```

Comment:
One thing that we are going to want to think about is how we'd do rolling upgrades. This is on the cards for next cycle.